_inc/client/components/number-format/index.js

```diff
@@ -0,0 +1,62 @@
+/**
+ * WordPress dependencies
+ */
+import { __experimentalGetSettings } from '@wordpress/date';
+
+/**
+ * Clean up WP locale so it matches the format expected by browsers.
+ *
+ * @param {string} locale - Locale given by WordPress.
+ *
+ * @returns {string} Browser-formatted locale.
+ */
+export const cleanLocale = locale => {
+	const regex = /^([a-z]{2,3})(_[a-zA-Z]*?)?(?:_.*)?$/i;
```

Comment:
> BCP 47, which is what the browser needs, has a limit of 8 characters on variants. If you try to pass it 9, it'll throw.

Cool, that's good to know. I've updated the second capturing group based on your suggestion in 3ea7f9edb64d8630a7f34d6afe37432319066603, and updated the tests to add more use-cases.

> I'm really thinking the best thing would be if we could do like WP does for HTML lang attributes and prefer __( 'html_lang_attribute' ) when it's set, as I mentioned earlier.

This brings up an interesting problem I hadn't fully considered before. Your suggestion would give us the site locale, which works in most scenarios. However, I think what we would really want to use in the Jetpack dashboard is the locale used by the current user viewing the page. This is what is returned in `@wordpress/date`'s `__experimentalGetSettings` (data from Core's `get_user_locale ()`). I've updated the comments and the function name to make that clearer in 82b32a43571667cdb8c11856609baecde7f969b3.

> I can't even find how to set it to test. My docker dev environment site has only "English (United States)" available

You should be able to install new languages and switch to them via Settings > General:

![image](https://user-images.githubusercontent.com/426388/95890224-c7dda280-0d83-11eb-9c19-c61430b25d67.png)

All installed languages are then available for use by all users on the site under Users > My profile:

![image](https://user-images.githubusercontent.com/426388/95890324-e8a5f800-0d83-11eb-982f-fb3e32fd1a5b.png)

Reply:
> Your suggestion would give us the site locale, which works in most scenarios. However, I think what we would really want to use in the Jetpack dashboard is the locale used by the current user viewing the page.

Is there no equivalent to `__()` that allows specifying a locale (so we could pass in the user's locale) instead of using the site's locale?

Reply:
> Is there no equivalent to `__()` that allows specifying a locale (so we could pass in the user's locale) instead of using the site's locale?

... Looks like not directly, but there is `switch_to_locale()` and `restore_previous_locale()`. I see a few places in the existing code that do something like
```
$switched_locale = switch_to_locale( get_user_locale( $user ) );

$something = __( '...' );

if ( $switched_locale ) {
    restore_previous_locale();
}
```

Reply:
> there is switch_to_locale() and restore_previous_locale(). 

Yes, that would be an option in PHP, although it would be more complicated here in the React codex.

At this point, I'm personally okay with the current implementation, it seems to work well with the different locale variations we've come across.

Reply:
What does React have to do with it? I don't see that React does anything with the i18n directly, it just uses strings passed from PHP. The `switch_to_locale()` and `restore_previous_locale()` would be in packages/blocks/wordpress/wp-includes/script-loader.php. Where you currently have
```php
$scripts->add_inline_script(
    'wp-date',
    sprintf(
        'wp.date.setSettings( %s );',
        wp_json_encode(
            array(
                'l10n'     => array(
                    'locale'        => get_user_locale(),
```
that might change to
```php
$switched = switch_to_locale( get_user_locale() );
$locale = __( 'html_lang_attribute' );
if ( $switched ) {
    restore_previous_locale();
}
if ( 'html_lang_attribute' === $locale || preg_match( '/[^a-zA-Z0-9-]/', $locale ) ) {
    $locale = str_replace( '_', '-', get_user_locale() );
}

$scripts->add_inline_script(
    'wp-date',
    sprintf(
        'wp.date.setSettings( %s );',
        wp_json_encode(
            array(
                'l10n'     => array(
                    'locale'        => $locale,
```

Reply:
Right, that's what I meant. This would require making changes to WordPress Core (such as in `wp-includes/script-loader.php` like you mentioned, which is not an option for us right now), or passing extra data from PHP to js on our end [here)(https://github.com/Automattic/jetpack/blob/ebed46ae922e3e31e879106c34a14881e0b81697/_inc/lib/admin-pages/class.jetpack-react-page.php#L196) to use later on. I don't believe this is worth it.

Reply:
Ah, yeah. I'm still not used to having most stuff in an upstream that's onerous to fix.